Sources/IntegerUtilities/GCD.swift

```diff
@@ -0,0 +1,42 @@
+//===--- GCD.swift --------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Numerics open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift Numerics project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// The greatest common divisor of `a` and `b`.
+///
+/// If both inputs are zero, the result is zero. If one input is zero, the
+/// result is the absolute value of the other input.
+///
+/// [wiki]: https://en.wikipedia.org/wiki/Greatest_common_divisor
+@inlinable
+public func gcd<T: BinaryInteger>(_ a: T, _ b: T) -> T {
+  var x = a.magnitude
+  var y = b.magnitude
+  
+  if y == 0 { return T(x) }
```

Comment:
We need to define what happens for `gcd(0, Int.min)`. The implementation here will trap attempting to construct `Int(.min.magnitude)`, which seems undesirable.

Three options:
1. Make the return type `T.Magnitude` instead of `T`. Always defined, but inconvenient in expressions.
2. Define `gcd(0, .min)`, `(.min, 0)` and `(.min, .min)` to return `.min`.
3. Trap.

I think I'm weakly in favor of (2), but open to convincing otherwise.

Reply:
Swift has consistently taken the view that integer types model the ring of integers, ℤ, and mathematical operations behave accordingly. If the result of an operation does not fit in the destination type, the program traps.

That’s why I wrote the code it this way: the GCD is well-defined and non-negative for all inputs, and if the result overflows there is a trap. This is no different from the existing subtraction operator, eg. `0 - .min`.

That would suggest `gcd` should stay as-is, and if a programmer wants to guarantee it won’t trap then they can either test for `.min` first, or work with `Magnitude` via:

```
let g = gcd(a.magnitude, b.magnitude)
```

We could also add documentation similar to that for [`SignedNumeric.negate()`](https://github.com/apple/swift/blob/a375ffebcf40e9f3392e297c24afd73c2aeb3623/stdlib/public/core/Integers.swift#L303-L319) regarding `.min` to warn of the trap, however it’s worth noting that the [`BinaryInteger` subtraction operator](https://github.com/apple/swift/blob/a375ffebcf40e9f3392e297c24afd73c2aeb3623/stdlib/public/core/Integers.swift#L832-L852) has nothing of the sort.

• • •

That said, I’m not totally opposed to (2), and I do recognize the value in being able to calculate first and then check for an edge-case result after, rather than preemptively needing to check for every possible input that could trap.

Still, it seems highly unusual that a greatest common divisor calculation would produce a negative result.

Reply:
That's fine; it should be documented and test cases added.

Reply:
Okay I added documentation and tests for the overflow behavior.

…wait, I think `XCTExpectFailure` has higher version requirements than this package. Is there a different way the tests should be written?

Reply:
I was considering bumping the supported version for other reasons anyway post-1.0, so let me investigate that a little bit further first.

Reply:
Is there a way to wrap the `expectFailure` tests in a conditional `#if` block?

Reply:
Let's remove them for now. Annoying, but we'll deal with it.

Reply:
Okay, I commented out the `ExpectFailure` tests, with a `TODO` to enable them when possible.

If you’d prefer I remove them entirely, just let me know.